lib/Parser/source.cpp

```diff
@@ -82,183 +58,68 @@ std::string LocateSourceFile(
   }
   for (const std::string &dir : searchPath) {
     std::string path{dir + '/' + name};
-    struct stat statbuf;
-    if (stat(path.c_str(), &statbuf) == 0 && !S_ISDIR(statbuf.st_mode)) {
+    bool isDir{false};
+    auto er = llvm::sys::fs::is_directory(path, isDir);
+    if (!er && !isDir) {
       return path;
     }
   }
   return name;
 }
 
-static std::size_t RemoveCarriageReturns(char *buffer, std::size_t bytes) {
-  std::size_t wrote{0};
-  char *p{buffer};
-  while (bytes > 0) {
-    void *vp{static_cast<void *>(p)};
-    void *crvp{std::memchr(vp, '\r', bytes)};
-    char *crcp{static_cast<char *>(crvp)};
-    if (!crcp) {
-      std::memmove(buffer + wrote, p, bytes);
-      wrote += bytes;
-      break;
-    }
-    std::size_t chunk = crcp - p;
-    std::memmove(buffer + wrote, p, chunk);
-    wrote += chunk;
-    p += chunk + 1;
-    bytes -= chunk + 1;
-  }
-  return wrote;
+std::size_t RemoveCarriageReturns(llvm::MutableArrayRef<char> buf) {
+  auto end = llvm::remove_if(buf, [](const char c) { return c == '\r'; });
```

Comment:
Do you know if the lambda is inlined?  If not, we should probably check the performance vs the original loop & memmove.

Reply:
The original loop actually has O(n^2) complexity if I am not reading it wrong, whereas the complexity of remove_if is O(n). Regarding the specific question though, the lambda does get inlined by both compilers I've tested it on.

Reply:
Why do you think that the loop is O(n**2)?  It makes one pass over the source, one iteration per `\r`, and moves each of the other bytes exactly once.  And the new version doesn't use `memchr`, which has been highly optimized for SIMD.  I'd like to see actual measured performance comparisons before signing off on this.

Reply:
Memmove is an O(n) function and is called inside an O(n) loop. In the worst case where every character is a carriage return n^2 operations will have been performed.

Reply:
O(#of lines) * O(average line length) == O(#of lines) * O(#total bytes / #lines) == O(#total bytes).

Reply:
Sorry, I don't follow. Consider the pathological case where every character in the input is a carriage return. We then call memmove once per character (which is n characters), and each memmove is O(n).

Reply:
And each of those `memmove()` calls will be moving zero bytes.

Reply:
If the memmove moves 0 bytes then the array remains the same and we haven't removed any carriage returns?

Reply:
If the memmove moves 0 bytes then the array remains the same and we haven't removed any carriage returns?

Reply:
The `memmove()` calls are moving the chunks of source bytes that lie *between* the `\r` characters.  If the whole buffer contains nothing but `\r` characters, those chunks are all empty.  The final size of the compacted buffer is the sum of the sizes of these chunks, so that would be zero.

Reply:
To add some data to the discussion:  here's a simple timing harness for testing three alternative implementations on a megabyte-sized source buffer with `\r` characters every 20 bytes (a guess at a reasonable mean line length, change it if you like).  I compiled this with g++ 9.2.0 and ran it on an old Haswell deskside machine with its Linux clock governor set to the `performance` mode.  The overhead of filling the source array 10,000 times is 0.86 total seconds.  The time needed to fill the source buffer and run the existing `RemoveCarriageReturns()` routine 10,000 times was 4.34 seconds.  The time for the `ByteAtATime()` alternative was 6.66 seconds.  The time for the predicate-based `remove_if()` alternative was 21.62 seconds.

```
#include <cstdlib>
#include <cstring>
#include <functional>
#include <iostream>

static std::size_t RemoveCarriageReturns(char *buffer, std::size_t bytes) {
  std::size_t wrote{0};
  char *p{buffer};
  while (bytes > 0) {
    void *vp{static_cast<void *>(p)};
    void *crvp{std::memchr(vp, '\r', bytes)};
    char *crcp{static_cast<char *>(crvp)};
    if (!crcp) {
      std::memmove(buffer + wrote, p, bytes);
      wrote += bytes;
      break;
    }
    std::size_t chunk = crcp - p;
    std::memmove(buffer + wrote, p, chunk);
    wrote += chunk;
    p += chunk + 1;
    bytes -= chunk + 1;
  }
  return wrote;
}

static std::size_t ByteAtATime(char *buffer, std::size_t bytes) {
  char *q{buffer};
  for (char *p{buffer}; bytes-- > 0; ++p) {
    if (*p != '\r') {
      *q++ = *p;
    }
  }
  return q - buffer;
}

static std::size_t remove_if(char *buffer, std::size_t bytes, const std::function<bool(char)> &predicate) {
  char *q{buffer};
  for (char *p{buffer}; bytes-- > 0; ++p) {
    if (!predicate(*p)) {
      *q++ = *p;
    }
  }
  return q - buffer;
}

int main() {
  int n{1 << 20};
  int linelen{20};
  char *src{static_cast<char *>(std::malloc(n))};
  int left;
  for (int reps{0}; reps < 10000; ++reps) {
    std::memset(src, 0, n);
    for (int j{0}; j < n; j += linelen) {
      src[j] = '\r';
    }
// Uncomment one of these three lines
//    int left = RemoveCarriageReturns(src, n);
//    int left = ByteAtATime(src, n);
//    int left = remove_if(src, n, [](char c){ return c == '\r'; });
    if (reps == 0) {
      std::cout << left << '\n';
    }
  }
  return 0;
}
```

Reply:
In order to find the next non - carriage return you have to have called another O(n) function though? 

Regardless, I was asked specifically by someone on llvm-dev to change the way character operations were done in this file to use C++ algorithms, which is why I made the change. On my system this function vectorises fine.
 I won't have time to do a thorough performance analysis before the merge and take that to llvm-dev to get consensus that the change doesn't need to be made. If we still think this is necessary can we do it as a post merge review? Otherwise we will need to delay the merge with llvm.

Reply:
Why have you written your own remove_if function? The one here implies 2 pointer indirections per predicate call (due to how std::function is implemented) and isn't inlineable. Neither of these things are true of std::remove_if. Your library implementation possibly also includes vectorisation hints

Reply:
Yes, I'm calling `memchr()` and `memmove()` in the loop.  Those are indeed O(n) functions.  But that does *not* make the overall function O(n**2), nor does it change the actual measured advantages of using optimized library routines in cases where performance matters.

I'm done here.

Reply:
std::remove_if is _also_ an optimised library routine, and you haven't measured the performance of that that I can see. Since that's the one I'm calling that's really the measurement we need to consider isn't it?

Reply:
Using `std::remove_if()` yields a time of 11.81 seconds, so it's about half the speed of the `ByteAtATime()` routine or a third the speed of the optimized routine.

Reply:
This change was as a prerequisite of the merge in this thread on llvm-dev: http://lists.llvm.org/pipermail/llvm-dev/2020-February/139458.html
If we aren't going to make the change for the reasons you've outlined would you possibly be able to take that discussion up on the list?

Reply:
I don't subscribe to any of these mailing lists; sorry.

Reply:
I can't replicate your performance results above; I've put my findings using each compiler and standard library combination I regularly test with below. 

I have only done a single run each time so the margin of error is larger than I would like; but in essence if we assume 0.5s to be our margin of error (which seems reasonable given the below results) then there is only one case when RemoveCarriageReturns is faster than std::remove_if, which seems to be a regression in libcxx9. 

On x86 macOS std::remove_if gives slightly faster performance than RemoveCarriageReturns, albeit not hugely outside my stated margin of error, and on Arm64 std::remove_if is around _40% faster_ than RemoveCarriageReturns when using libstdcxx, and within the margin of error when using libcxx.

These results were obtained using your test program with the additional `remove_if` line you gave, and in each case I only used the flags `-std=c++17` and `-O3`. A quick test of a couple of them with `-O2` demonstrated that there was no difference between the two optimisation levels, so I continued to use the one that LLVM and the reworked CMake for LLVM use.

# Arm64 Linux
## Clang 9 libstdcxx 9
`std::remove_if` - 0m7.606s
`RemoveCarriageReturns` - 0m12.148s

## Clang 9 libcxx 9
`std::remove_if` - 0m14.480s
`RemoveCarriageReturns` - 0m12.239s

## Clang 8 libstdcxx 9
`std::remove_if` - 0m7.563s
`RemoveCarriageReturns` - 0m12.219s

## Clang 8 libcxx 8
`std::remove_if` - 0m12.134s
`RemoveCarriageReturns` - 0m12.103s

## Clang 7 libstdcxx 9
`std::remove_if` - 0m7.504s
`RemoveCarriageReturns` - 0m12.097s

## GCC 9
`std::remove_if` - 0m6.894s
`RemoveCarriageReturns` - 0m12.701s

## GCC 8
`std::remove_if` - 0m6.955s
`RemoveCarriageReturns` - 0m12.646s

## GCC 7
`std::remove_if` - 0m6.946s
`RemoveCarriageReturns` - 0m12.585s

# x86-64 macOS
## Clang 9 libcxx 9
`std::remove_if` - 0m3.316s
`RemoveCarriageReturns` - 0m4.629

## Clang 8 libcxx 8
`std::remove_if` - 0m3.641s
`RemoveCarriageReturns` - 0m4.873s

## Clang 7 libcxx 7
`std::remove_if` - 0m4.345s
`RemoveCarriageReturns` - 0m4.836s

## Apple Clang 11
`std::remove_if` - 0m4.454s
`RemoveCarriageReturns` - 0m4.413s

Reply:
And were you able to reproduce my results (gcc 9.2.0, -O3, x86, cpufreq governor set to `performance`)?

Reply:
What were the times from the `ByteAtATime()` alternative in these environments?

Reply:
I cannot replicate your results on X86 Linux with gcc and performance governor either, see below.
I didn't test the other implementation as it isn't being proposed here.

# x86_64 Linux
## GCC 9
`std::remove_if` - 0m3.363s
`RemoveCarriageReturns` - 0m4.775s

## GCC 8
`std::remove_if` - 0m3.388s
`RemoveCarriageReturns` - 0m7.851s

## GCC 7
`std::remove_if` - 0m5.688s
`RemoveCarriageReturns` - 0m8.083s

Reply:
The reason to measure the `ByteAtAtTime()` alternative is that it's a sanity check on the relationship between the other measurements.

Reply:
I have added results for `ByteAtATime` above

Reply:
As a sanity check I think that ByteAtATime looks odd in your results; ByteAtATime is effectively the “simple” implementation of std::remove_if, if that function had no optimisations in the library. So remove_if should be at least as fast as ByteAtATime, as the performance of ByteAtATime is what remove_if would have if it had no optimisations in the library implementation.

Reply:
I have managed to somewhat replicate your results on my home desktop. Using a timer that only measures the loops, increasing the number of iterations to get clearer results and with the times given in microseconds I get;
`std::remove_if` - 698210 average per iteration
`RemoveCarriageReturns` - 503866
`ByteAtATime` - 693623

So I don't see any significant difference between `ByteAtATime` and `std::remove_if` still, but `RemoveCarriageReturns` is faster on this specific setup.

Reply:
How can we move forward here?

Clearly we don't want to degrade performance unnecessarily, but that is not the only concern. The coding style of the proposed new implementation is certainly closer to what the LLVM community would expect and this particular code has been requested to be re-written - http://lists.llvm.org/pipermail/llvm-dev/2020-February/139464.html - before we upstream.

Given the data shown, I don't think it is clear that David's proposed new implementation is systematically worse than the current implementation. It seems that results vary depending probably on how optimised your C and C++ standard libraries are for the system you are on,

If we agree on that, I think we should make the proposed change on the grounds of coding style alignment to LLVM to unblock upstreaming.

Reply:
Dealing with DOS line endings is something that we have to do most often on Windows than elsewhere, so the x86 timings seem to indicate to me that we'll want to retain the fast approach in NVIDIA's product sources.  You can do whatever you think you have to in your LLVM fork.

Reply:
I think in order to make any statements about windows we would have to test there rather than Linux; their C and C++ standard library implementations are completely different to the ones on Linux and therefore very likely to have different performance characteristics. In addition my macOS results still show `remove_if` as being faster than the hand rolled function here, and that's also on x86. Again I suspect that this is because the C standard library on macOS is different to on Linux.

Reply:
This function does also have to run on every platform, regardless of whether the input actually has carriage returns or not. So the performance matters everywhere not just on windows on x86.

Reply:
@sscalpone 
So that we can progress the rest of the patch, David has removed this change from the review. Can we merge as-is?